### PR TITLE
Redo segmentations

### PR DESCRIPTION
**Pull request by MikeLippincott**
This PR performs segmentations within python to import masks into CellProfiler later for feature extraction

### CODE REVIEW

File: Wave1_data/3.cell_segmentation/figures/cell_segmentation_differences.png
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/figures/cell_segmentation_differences.png:1]: Woahhh that IC hurts haha. Maybe would require some changes, one I can recommend is definitely changing from background to regular method and go from there. Very odd!

Also GSDMC is definitely better for segmentation only because there is clear contrast, those other channels are very messy... interesting stuff!

File: Wave1_data/3.cell_segmentation/figures/unique_cell_masks_vs_diameter.png
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/figures/unique_cell_masks_vs_diameter.png:1]: Looks to be empty unfortunately :(((((

File: Wave1_data/3.cell_segmentation/figures/unique_nuclei_masks_vs_diameter.png
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/figures/unique_nuclei_masks_vs_diameter.png:1]: What is the red line? Mean of the diameter?

Also, very interesting how small the nuclei are, but I am not as familiar with the data so it might be expected! Might need some coSMicQC if those diameters are cluster nuclei/over-segmented haha

File: Wave1_data/3.cell_segmentation/notebooks/0.nuclei_segmentation.ipynb
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/notebooks/0.nuclei_segmentation.ipynb:1]: 45 diameter looks to be pretty optimal! Unfortunately I can't see all of the options though.

One main thing I noticed (that we chatted about in person) was that these nuclei look very messy and look to contain mycoplasma, in my opinion. I have only seen it once and these nuclei look very similar to what I saw for mycoplasma.

As we discussed, it would be good to ask the collaborators to test the cells to ensure no contamination.

File: Wave1_data/3.cell_segmentation/notebooks/1.cell_segmentation.ipynb
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/notebooks/1.cell_segmentation.ipynb:1]: I don't see any examples in this file for what the cell segmentations look like, so I can't provide feedback on that unfortunately :(((

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# This notebook focuses on trying to find a way to segment cells within organoids properly.
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]:
```suggestion
# This notebook focuses on trying to find a way to segment nuclei properly.
```
I think this code comment is copied from a different project, so recommend correcting.

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+# The end goals is to segment cell and extract morphology features from cellprofiler.
+# These masks must be imported into cellprofiler to extract features.
+
+# In[ ]:
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]: Looks like this is missing running the first code cell.

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+except NameError:
+    in_notebook = False
+
+if in_notebook:
+    print("Running as a notebook")
+elif not in_notebook:
+    print("Running as script")
+
+
+# In[ ]:
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]: Also missing here.

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    plt.ylabel("Number of unique masks")
+    plt.title("Number of unique masks vs Diameter")
+    if in_notebook:
+        plt.show()
+    plt.savefig("../figures/unique_nuclei_masks_vs_diameter.png")
+
+
+# ### Runnning segmentation
+
+# In[ ]:
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]: Also missing here.

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            plt.title(f"Diameter: {diameters[diameter]}")
+            plt.axis("off")
+            plt.subplot(1, 2, 1)
+            plt.imshow(imgs[diameter], cmap="gray")
+            plt.title("Nuclei")
+            plt.axis("off")
+
+            plt.subplot(122)
+            plt.imshow(render_label(masks_all[diameter]))
+            plt.title("Cell masks")
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]:
```suggestion
            plt.title("Nuclei masks")
```

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+    channels = [[1, 0]]
+
+    masks_all_dict = {"masks": [], "imgs": []}
+
+    # get masks for all the images
+    # save to a dict for later use
+    for img in nuclei:
+        img = normalize(img)
+        masks, flows, styles, diams = model.eval(img, channels=channels, diameter=50)
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py:234]: Totally agree with this diameter!

File: Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# This notebook focuses on trying to find a way to segment cells within organoids properly.
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py]:
```suggestion
# This notebook focuses on trying to find a way to segment cells properly.
```
Same comment as in the nuclei segmentation script.

File: Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    RGB = (RGB / RGB.max() * 255).astype(np.uint8)
+
+    rgb_image_pil = Image.fromarray(RGB)
+
+    imgs.append(rgb_image_pil)
+
+
+# ## Cellpose
+
+# In[ ]:
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py]: Missing here.

File: Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+
+    channels = [[1, 3]]  # channels=[red cells, blue nuclei]
+
+    masks_all_dict = {"masks": [], "imgs": []}
+    imgs = np.array(imgs)
+
+    # get masks for all the images
+    # save to a dict for later use
+    for img in imgs:
+        # masks, flows, styles, diams = model.eval(img, diameter=diameter, channels=channels)
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py]: Recommend removing commented out code.

File: Wave1_data/3.cell_segmentation/scripts/2.cytoplasm_issue_visualization.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# This notebook focuses on trying to find a way to segment cells within organoids properly.
+# The end goals is to segment cell and extract morphology features from cellprofiler.
+# These masks must be imported into cellprofiler to extract features.
```
Comment (jenna-tomkinson) [inline Wave1_data/3.cell_segmentation/scripts/2.cytoplasm_issue_visualization.py]: Consider updating this to reflect the cytoplasm issue checking.

**Review by jenna-tomkinson — APPROVED** [review]
LGTM! Exciting to see this optimization, super interesting! I left some comment and one major concern for mycoplasma with the nuclei, other than that nice job!

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py (hunk 2 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]: reran

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py (hunk 3 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]: reran

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py (hunk 4 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py]: reran

File: Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py (hunk 6 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/0.nuclei_segmentation.py:234]: :)

File: Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py (hunk 8 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py]: reran

File: Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py (hunk 9 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/1.cell_segmentation.py]: removed!

File: Wave1_data/3.cell_segmentation/scripts/2.cytoplasm_issue_visualization.py (hunk 10 above)
Comment (MikeLippincott) [reply Wave1_data/3.cell_segmentation/scripts/2.cytoplasm_issue_visualization.py]: changed!
